addons/account/models/account_move.py

```diff
@@ -816,7 +816,12 @@ def _compute_highest_name(self):
 
     @api.model
     def _get_query_made_hole(self, ids=None):
-        return f"""
+        ids_domain = SQL()
+        if ids:
+            ids_domain = SQL("AND this.id = ANY(%s)", ids)
+        elif irregular_domain := self.env.context.get('irregular_domain'):
+            ids_domain = SQL("AND this.id IN (%s)", self._where_calc(irregular_domain).select())
```

Comment:
It can be a combination of multiple prefixes, there is no reason to not be as specific as possible

Reply:
With the following quick ugly diff (missing filtering on company_id too, but it's accounted for in the dalibo), there was a ~30% time decrease for our own prod. Which I think is due to the removal of subselect. If we know the fields that we want to use to filter, I don't understand why not using them directly instead of subquerying

https://explain.dalibo.com/plan/f55g4h85157f6287#plan
https://explain.dalibo.com/plan/e25g4g3cb0036494#plan

```diff
diff --git a/addons/account/models/account_journal_dashboard.py b/addons/account/models/account_journal_dashboard.py
index 96594c9f2871..95294904e73f 100644
--- a/addons/account/models/account_journal_dashboard.py
+++ b/addons/account/models/account_journal_dashboard.py
@@ -917,6 +917,8 @@ class account_journal(models.Model):
             ]
             for journal_id, prefix in has_sequence_holes
         )
+        sequence_prefixes = tuple([prefix for __, prefix in has_sequence_holes])
+        journal_ids = tuple([journal_id for journal_id, __ in has_sequence_holes])
         return {
             'type': 'ir.actions.act_window',
             'name': _("Journal Entries"),
@@ -930,6 +932,8 @@ class account_journal(models.Model):
                 'search_default_irregular_sequences': 1,
                 'expand': 1,
                 'irregular_domain': to_tuple(domain),
+                'sequence_prefixes': sequence_prefixes,
+                'journal_ids': journal_ids,
             }
         }
 
diff --git a/addons/account/models/account_move.py b/addons/account/models/account_move.py
index cea24d7487aa..e32e7c76bfec 100644
--- a/addons/account/models/account_move.py
+++ b/addons/account/models/account_move.py
@@ -816,12 +816,16 @@ class AccountMove(models.Model):
 
     @api.model
     def _get_query_made_hole(self, ids=None):
-        ids_domain = SQL()
+        extra_condition = SQL()
         if ids:
-            ids_domain = SQL("AND this.id = ANY(%s)", ids)
-        elif irregular_domain := self.env.context.get('irregular_domain'):
-            ids_domain = SQL("AND this.id IN (%s)", self._where_calc(irregular_domain).select())
-        return SQL("""
+            extra_condition = SQL("AND this.id = ANY(%s)", ids)
+        elif self.env.context.get('irregular_domain'):
+            extra_condition = SQL(
+                "AND this.sequence_prefix IN %s AND this.journal_id IN %s",
+                tuple(self.env.context.get('sequence_prefixes')),
+                tuple(self.env.context.get('journal_ids')),
+            )
+        sql = SQL("""
                 SELECT this.id
                   FROM account_move this
                   JOIN res_company company ON company.id = this.company_id
@@ -832,7 +836,8 @@ class AccountMove(models.Model):
                    AND this.sequence_number != 1
                    AND this.name != '/'
                    %s
-        """, ids_domain)
+        """, extra_condition)
+        return sql
 
     @api.depends('name', 'journal_id')
     def _compute_made_sequence_hole(self):

```

Reply:
But the resulting query should be exactly the same?
And the plans are not comparing your version of the query with mine

Reply:
Also the explain you did don't contain the domain hardcodes in the action of the view, which is pretty important to compare both versions

Reply:
> But the resulting query should be exactly the same? And the plans are not comparing your version of the query with mine

What do you mean the same ? The queries are different :thinking: 

```sql
 SELECT this.id
  FROM account_move this
  JOIN res_company company ON company.id = this.company_id
LEFT JOIN account_move other ON this.journal_id = other.journal_id
                          AND this.sequence_prefix = other.sequence_prefix
                          AND this.sequence_number = other.sequence_number + 1
  WHERE other.id IS NULL
    AND this.sequence_number != 1
    AND this.name != '/'
    AND this.sequence_prefix IN ('INV/2024/')
    AND this.journal_id IN (1)
    AND this.company_id IN (1)
```

vs

```sql
SELECT this.id
  FROM account_move this
  JOIN res_company company ON company.id = this.company_id
LEFT JOIN account_move other ON this.journal_id = other.journal_id
                          AND this.sequence_prefix = other.sequence_prefix
                          AND this.sequence_number = other.sequence_number + 1
  WHERE other.id IS NULL
    AND this.sequence_number != 1
    AND this.name != '/'
    AND this.id IN (
      SELECT "account_move"."id"
      FROM "account_move"
      WHERE (((
        "account_move"."company_id" IS NULL 
        OR ("account_move"."company_id" IN (1))) 
        AND ("account_move"."journal_id" = 1)) 
        AND ("account_move"."sequence_prefix" = 'INV/2024/')))
```

But in the end, going from 10s to 5ms or 4ms isn't that much different tbh and keeping the subselect may be less annoying if we ever update the domain in the future. So let's go with this then :)